### DISCUSSION TITLE
Is BeanDeserializerModifier expected to work for immutable Beans?

### DISCUSSION THREAD

**Discussion by herkrath** [FasterXML/jackson-databind#4351] [General]
Hi @cowtowncoder, all,

Is BeanDeserializerModifier expected to work for immutable Beans? It is not documented otherwise and the BeanDeserializerModifier methods are triggered, so I would assume yes. It doesn't work though. I tested with a mutable (working) and immutable (not working) version of the same Bean with Jackson 2.14 and 2.16.1.

This is not a recently introduced bug, there are multiple older discussions about this topic, e.g. https://stackoverflow.com/questions/54086246/jacksons-beandeserializermodifier-does-not-work-with-final-fields

The suggested solution in there is not always applicable but the reference to BeanDeserializerBase::resolve seems correct. It appears to me that creator properties aren't correctly set based on BeanDeserializerModifier modifications.

I am mainly looking for a confirmation.

Thanks in advance,
Mark

**Comment by cowtowncoder** [comment]
@herkrath At high level, there is nothing specific wrt mutable/immutable beans wrt `BeanDeserializerModifier`, as it only operates on definition of POJO, before/during construction of `JsonDeserializer`. So it should be fully usable in that sense.

If there are problems in actually using it, those may be flaws or limitations, but there is no intent to prevent its use. There are some known problems specifically related to Creator properties of "implicit" Creators (ones that are not annotated with `@JsonCreator`), in which annotations are not properly merge across Creator and other properties. So this could be what is happening. But that is sort of outside control of `BeanDeserializerModifier`.

There are plans to try to tackle this problem wrt (some) Creator properties: as a pre-cursor, I am working on #3719 . But full introspection rewrite has been postponed multiple times as it is likely tricky to do. Even if conceptually simple: introspection should start with location Creators, then proceed with other properties.
But existing processing is complicated so it is not necessarily trivial to change this.

**Reply by herkrath** [reply]
It also occurs while having `@JsonCreator` in place. Shall I create an issue for it, containing some samples? I can also try to provide a PR but as you say "existing processing is complicated", so I am not sure that I wouldn't break anything else. But if it seems helpful to you, I will try.

**Reply by cowtowncoder** [reply]
If you can distill it down to something that can show the problem, sure. There are quite a few issues about various aspects of the problem, so not a big deal if you can't extract a test case, but if you can that could be useful.

**Reply by herkrath** [reply]
I have written a unit test which is showing the issue. While doing so, I realized that I should make one point more clear: `BeanDeserializerModifier` seems to work with immutable beans in general (i.e. `modify*Deserializer` methods). What does not work is custom `updateBuilder` and possibly `updateProperties` logic for the properties being using inside the bean.

I will try to fix this next.

**Reply by cowtowncoder** [reply]
Right, it is possible some aspects won't work well.

**Reply by herkrath** [reply]
Hi @cowtowncoder, I have created the issue #4356 and the PR #4357 for the point about the serializer. Sorry if I spam your repo, I am not used to GitHub. Should I have simply created the PR without an issue?

I also created failing UTs about another related issue: Changing the property order via `BeanDeserializerModifier::updateProperties` doesn't seem to work at all, not for mutable, nor for immutable beans. How to pass them best (here, via issue with GitHub link, via PR)?

**Reply by cowtowncoder** [reply]
No, creating an Issue and PR is pretty standard (although sometimes just PR is fine too; both are fine).

Failing test (under `src/test/java/.../failing`) would be a good PR to show the problem.
Although I am surprised as I think `updateProperties()` is definitely used somewhere. But maybe test can show what is going on.

**Reply by herkrath** [reply]
Alright, `BeanDeserializerModifier::updateProperties` issue is #4358 and PR #4359 (failing UTs only).

**Reply by cowtowncoder** [reply]
Finally found time to dig deep and was able to fix #4356 (for 2.18.0 release). Will now look into #4358 to see what gives there.
